src/coreclr/System.Private.CoreLib/src/System/Threading/Thread.CoreCLR.cs

```diff
@@ -206,19 +222,32 @@ public bool IsBackground
             }
         }
 
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        private extern bool GetIsBackground();
+        [SuppressGCTransition]
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "ThreadNative_GetIsBackground")]
+        private static partial Interop.BOOL GetIsBackground(ThreadHandle t);
 
         [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "ThreadNative_SetIsBackground")]
         private static partial void SetIsBackground(ThreadHandle t, Interop.BOOL value);
 
         /// <summary>Returns true if the thread is a threadpool thread.</summary>
-        public extern bool IsThreadPoolThread
+        public bool IsThreadPoolThread
         {
-            [MethodImpl(MethodImplOptions.InternalCall)]
-            get;
-            [MethodImpl(MethodImplOptions.InternalCall)]
-            internal set;
+            get
+            {
+                if (_isDead)
+                {
+                    throw new ThreadStateException(SR.ThreadState_Dead_State);
+                }
+
+                return _isThreadPool;
+            }
+            internal set
+            {
+                Debug.Assert(value);
+                Debug.Assert(!_isDead);
+                Debug.Assert(_isThreadPool || ((ThreadState & ThreadState.Unstarted) != 0));
```

Comment:
```suggestion
                Debug.Assert((ThreadState & ThreadState.Unstarted) != 0) || ThreadPool.UseWindowsThreadPool);
```
I think this is the right way to fix this assert. (Alternative is to delete it.)